### PR TITLE
feat: HSM support

### PR DESCRIPTION
**Pull request by bgrieder**
The main goal of this PR is to add support for the Proteccio HSM.
It provides both
 - the ability to perform the Create, Destroy, Export, Encrypt, and Decrypt operations on the HSM
 - the ability to create keys in the KMS which are wrapped by a key in the HSM
 
 This leads to significant refactoring of the database and encryption/decryption (wrap/unwrap) components to add flexibility for supporting additional hardware (or external software) in the future. In particular:
 - the database components are now in a separate crate `server_database`. They are now split in 2 implementations: Objects store and Permissions store
 - a new `interfaces` crate gathers interfaces to be implemented by new external components. Interfaces include:
     - Object Store
     - Permissions Store
     - Encryption Oracle
     
Key unique identifiers now support prefixes. Object Stores, Permissions stores, and Encryption Oracles can be registered against the prefixes. 

Support for the `Sensitive` Attribute in addition to the ability to wrap a key by another key has been added to all keys creations

### CODE REVIEW

**Review by Manuthor — COMMENTED** [review]
review in progress

File: crate/cli/src/tests/hsm/encrypt_decrypt.rs [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+        symmetric::{create_key::create_symmetric_key, encrypt_decrypt::run_encrypt_decrypt_test},
+    },
+};
+
+#[test]
+pub(crate) fn test_aes_gcm() -> CliResult<()> {
+    let dek = create_symmetric_key(
+        KMS_HSM_CLIENT_CONF,
+        CreateKeyAction {
+            key_id: Some("hsm::4::".to_string() + &Uuid::new_v4().to_string()),
```
[the remark this answers is not in the document]
Comment (Manuthor) [reply crate/cli/src/tests/hsm/encrypt_decrypt.rs:32]: (OK it is the slot number)

[the remark this answers is not in the document]
Comment (bgrieder) [reply crate/cli/src/tests/hsm/encrypt_decrypt.rs:32]: Yes

[the remark this answers is not in the document]
Comment (bgrieder) [reply crate/cli/src/tests/hsm/encrypt_decrypt.rs:32]: Should be explicit in the public documentation

File: crate/cli/src/tests/hsm/mod.rs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+//! ```
+
+#[cfg(not(feature = "fips"))]
+use crate::tests::hsm::encrypt_decrypt::test_rsa_pkcs_v15;
+use crate::{
+    error::result::CliResult,
+    tests::hsm::encrypt_decrypt::{test_aes_gcm, test_rsa_pkcs_oaep},
+};
+
+const KMS_HSM_CLIENT_CONF: &str = "/home/bgrieder/.cosmian/kms_hsm.json";
```
[the remark this answers is not in the document]
Comment (bgrieder) [reply crate/cli/src/tests/hsm/mod.rs:16]: I will fix this

File: crate/cli/src/tests/cover_crypt/encrypt_decrypt.rs [hunk 3]
```diff
@@ -339,30 +347,34 @@ async fn test_encrypt_decrypt_using_tags() -> CliResult<()> {
...
-            &ctx.owner_client_conf_path,
-            &[output_file.to_str().unwrap()],
-            "[\"tag\"]",
-            Some(recovered_file.to_str().unwrap()),
-            Some("myid"),
-        )
-        .is_err()
-    );
+    //TODO Left here but this has become undefined behavior in the new version:
+    //TODO if the first key found is the correct one, decryption will work, else it will fail
```
[the remark this answers is not in the document]
Comment (bgrieder) [reply crate/cli/src/tests/cover_crypt/encrypt_decrypt.rs:351]: I think it is fine this way: if multiple keys match a user query, there is no way to determine which key is correct. Going through all the keys to discover multiple keys match and warning the user is too costly, or not even practically feasible on an HSM

**Pull request merged**